pkg/upgrade/upgrades/update_invalid_column_ids_in_sequence_back_references.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package upgrades
+
+import (
+	"context"
+	"sort"
+
+	"github.com/cockroachdb/cockroach/pkg/clusterversion"
+	"github.com/cockroachdb/cockroach/pkg/jobs"
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/upgrade"
+	"github.com/cockroachdb/errors"
+)
+
+func updateInvalidColumnIDsInSequenceBackReferences(
+	ctx context.Context, version clusterversion.ClusterVersion, d upgrade.TenantDeps, job *jobs.Job,
+) error {
+	var lastUpgradedID descpb.ID
+	// Upgrade each sequence, one at a time, until we exhaust all of them.
+	for {
+		done, idToUpgrade, err := findNextTableToUpgrade(ctx, d.InternalExecutor, lastUpgradedID,
```

Comment:
Does it make sense to have both `findNextTableToUpgrade` and `maybeUpdateInvalidColumnIdsInSequenceBackReferencesBestEffort` encompassed in a `d.CollectionFactory.Txn()`?

Reply:
What difference will that make? Will there be any correctness concern the way this is currently implemented? 

We used `tenantDeps.InternalExecutor` to retrieve all the sequence IDs in the cluster, and then for each of them, we used `tenantDeps.CollectionFactory.Txn` to perform an upgrade. Will there be any scenario where this pattern of usage might yield incorrect results?

Reply:
Could the descriptor we obtained from `findNextTableOrViewToUpgrade` be dropped by another goroutine before we execute the upgrade? 
Sorry that I'm not 100% sure if my concern is valid -- please take it as an open question rather than a change request.

Reply:
I see your point now -- `collectionFactory.TxnWithExecutor` nicely gives us an internal executor that is tied to the collection. I have changed the code as you suggested.

Reply:
Yes, the descriptor could have been modified in parallel. The good news, I think, is that the logic here is monotone. If there was a parallel modification, we'll see it when we retrieve the descriptor and we'll notice that no upgrade needs to happen. Do you see some hazard I'm missing?

Reply:
I see, thanks for the explanation! That's the only hazard I could think of here.